src/tidb-kv-client.md

```diff
@@ -0,0 +1,86 @@
+# TiDB KV Client
+
+## TODOs
+
+- [ ] Data Source Executors
+- [ ] DistSQL: ...
+- [ ] Cop Client: ...
+- [ ] Txn KV: Snapshot Read
+  - [ ] Get
+  - [ ] BatchGet
+  - [ ] Iter
+  - [ ] IterReverse
+- [ ] Txn KV: Lock Keys
+- [ ] Txn KV: Commit
+- [ ] Txn KV: Rollback
+
+## RPC Client: Send Request
+
+```railroad
+Diagram(
+  NonTerminal("Get conn pool to the target store"),
+  Choice(
+    0,
+    Sequence(
+      Comment("Batch enabled"),
+      NonTerminal("Push request to channel", {cls: "with-metrics"}),
+      NonTerminal("Wait response", {href: "#rpc-client-batch-request-loop"}),
+    ),
+    Sequence(
+      NonTerminal("Get conn from pool"),
+      NonTerminal("Call RPC"),
+      Choice(
+        0,
+        Comment("Unary call"),
+        NonTerminal("Recv first"),
+      ),
+    ),
+  ),
+)
+```
+
+- The overall duration of sending a request is obeserved as `tidb_tikvclient_request_seconds`.
+- RPC client maintains connection pools (named *ConnArray*) to each store, and each pool has a *BatchConn* with a [batch request(send) channal](#rpc-client-batch-request-loop).
+- Batch is enabled when the store is tikv and batch size is positive, which is true in most cases.
+- The size of batch request channel is `tikv-client.max-batch-size` (default: 128), the duration of enqueue is obeserved as `tidb_tikvclient_batch_wait_duration`.
+- There are three kinds of stream request: CmdBatchCop, CmdCopStream, CmdMPPConn, which involve an additional recv call to fetch the first response from the stream.
+
+## RPC Client: Batch Request Loop
+
+```railroad
+Diagram(
+  Comment("send"),
+  OneOrMore(
+    Sequence(
+      NonTerminal("Fetch pending requests"),
+      Choice(
+        0,
+        Comment("tikv is not overload"),
+        NonTerminal("Fetch more requests"),
+      ),
+      NonTerminal("Select a connection"),
+      NonTerminal("Build a batch request and send it"),
```

Comment:
I think this step only appends the batch request into the control buffer of the streaming connection and another goroutine will send it asynchronously. So there should be a missing latency after this step.

Reply:
The diagram describes the [batch send loop](https://github.com/tikv/client-go/blob/be31f33ba03b543bb20503c3264877e09440d30c/internal/client/client_batch.go#L299) and this step is corresponding to [here](https://github.com/tikv/client-go/blob/be31f33ba03b543bb20503c3264877e09440d30c/internal/client/client_batch.go#L374-L385). Do you mean https://github.com/you06/performance-map/pull/3/files#diff-8354e42366dea0e62bb1095703d36c40ada5924ddd9c1d817ff2e5736b7ededbR26-R27 ?

Reply:
What I mentioned is the mechanism inside gRPC, it has a [buffer](https://github.com/grpc/grpc-go/blob/71f16a1d4afab21dea1732c89401d2e7b9949d61/internal/transport/http2_client.go#L79-L81) for messages to be sent. The client just [put the messages into the buffer](https://github.com/grpc/grpc-go/blob/71f16a1d4afab21dea1732c89401d2e7b9949d61/internal/transport/http2_client.go#L974) by calling `Send`.

Reply:
I see, how can we observe such a duration?

Reply:
It's unable to observe such duration unless we can add some metrics inside gRPC. Let's left a grey block for the unclear part.

Reply:
What about

```
NonTerminal("Build a batch request and append it to the sending buffer"),
NonTerminal("Send asynchronously"),
```

The first section is under client-go's control and the second section can be related to the utilization of the system.